File: homework_12/resources/BaseFileHandler.java
package resources;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.io.BufferedReader;
import java.io.BufferedWriter;

public abstract class BaseFileHandler {
    private File fileStorage;
    
    public BaseFileHandler(String argFilePath) throws IOException{
        this.fileStorage = new File(argFilePath);
        if (! this.fileStorage.exists()) {
            this.fileStorage.createNewFile();
        }
    }

    public void SetFile(File argIn){
        this.fileStorage = argIn;
    }

    public String Read(int argLineNumber) throws IOException{
        BufferedReader fileStream = new BufferedReader(new FileReader(this.fileStorage));
        String line;
        List<String> lines = (fileStream.lines().toList());
        if (argLineNumber - 1 < lines.size() && argLineNumber - 1 >= 0) {
            line = lines.get(argLineNumber - 1);
        } else {
            line = null;
        }
        fileStream.close();
        return line;
    }

    public void Edit(int argLineNumber, String argLine) throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        BufferedWriter fileStream = new BufferedWriter(new FileWriter(this.fileStorage));
        for (int i = 0; i < lines.size(); i++) {
            if (i + 1 == argLineNumber) {
                fileStream.append(argLine);
                fileStream.newLine();
            }
            else {
                fileStream.append(lines.get(i));
                fileStream.newLine();
            }
        }
        fileStream.flush();
        fileStream.close();
        tempReader.close();
    }

    public void Insert(int argLineNumber, String argLine) throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        BufferedWriter fileStream = new BufferedWriter(new FileWriter(this.fileStorage));
        for (int i = 0; i < lines.size(); i++) {
            if (i + 1 == argLineNumber) {
                fileStream.append(lines.get(i));
                fileStream.newLine();
                fileStream.append(argLine);
                fileStream.newLine();
            }
            else {
                fileStream.append(lines.get(i));
                fileStream.newLine();
            }
        }
        fileStream.flush();
        fileStream.close();
        tempReader.close();
    }

    public void Delete(int argLineNumber) throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        BufferedWriter fileStream = new BufferedWriter(new FileWriter(this.fileStorage));
        for (int i = 0; i < lines.size(); i++) {
            if (i + 1 != argLineNumber) {
                fileStream.append(lines.get(i));
                fileStream.newLine();
            }
        }
        fileStream.flush();
        fileStream.close();
        tempReader.close();
    }

    public void Write(String argLine) throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        BufferedWriter fileStream = new BufferedWriter(new FileWriter(this.fileStorage));
        for (String line : lines) {
            fileStream.append(line);
            fileStream.newLine();
        }
        fileStream.write(argLine);
        fileStream.newLine();
        fileStream.flush();
        fileStream.close();
        tempReader.close();
    }
    
    public void Clear() throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        BufferedWriter fileStream = new BufferedWriter(new FileWriter(this.fileStorage));
        for (int i = 0; i < lines.size(); i++) {
            if (i + 1 == 1 || i + 1 == lines.size()) {
                fileStream.append(lines.get(i));
                fileStream.newLine();
            }
        }
        fileStream.flush();
        fileStream.close();
        tempReader.close();
    }

    public int Find(String argLine) throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        int tempOut = -1;
        for (String line : lines) {
            if (line.contains(argLine)) tempOut = lines.indexOf(line);
        }
        tempReader.close();
        return tempOut;
    }

    public int Size() throws IOException{
        BufferedReader tempReader = new BufferedReader(new FileReader(this.fileStorage));
        List<String> lines = (tempReader.lines().toList());
        tempReader.close();
        return lines.size();
    }

    public void Copy(BaseFileHandler argIn) throws IOException{
        this.Clear();
        for (int i = 0; i < argIn.Size(); i++) {
            this.Insert(this.Size() - 1, this.formData(argIn.extractData(argIn.Read(i))));
         }
    }

    abstract public String formData(GameData argIn);
    
    abstract public GameData extractData(String argIn);
}

File: homework_13/sharp_project/BaseFileHandler.cs
using System.IO;
using System.Text;
public abstract class BaseFileHandler {
    private string info;
    private string filePath;

    private string fileStart;

    private string fileEnd;

    
    public BaseFileHandler(string argFilePath, string argFileStart, string argFileEnd, string argInfo) {  
        this.info = argInfo;
        this.filePath = argFilePath;
        this.fileStart = argFileStart;
        this.fileEnd = argFileEnd;
        if (! File.Exists(argFilePath)) {
            FileStream fileStorage = File.Create(argFilePath);
            fileStorage.Close();
            this.Clear();
        } 
    }

    public string Info(){
        return this.info;
    }

    public string Read(int argLineNumber){       
        string tempLine;
        string[] tempLines = File.ReadAllLines(this.filePath);
        tempLine = tempLines[argLineNumber -1];
        return tempLine;
    }

    public void Edit(int argLineNumber, string argLine){
        string[] tempLines = File.ReadAllLines(this.filePath);
        tempLines[argLineNumber - 1] = argLine;
        StringBuilder builder = new StringBuilder();
        foreach (string tempLine in tempLines) {
            builder.Append(String.Format("{0}\n", tempLine));
        }
        File.WriteAllText(this.filePath, builder.ToString());
    }

    public void Insert(int argLineNumber, string argLine){
        string[] tempLines = File.ReadAllLines(this.filePath);
        List<string> outLines = tempLines.ToList();
        outLines.Insert(argLineNumber - 1, argLine);
        StringBuilder builder = new StringBuilder();
        foreach (string tempLine in outLines) {
            builder.Append(String.Format("{0}\n", tempLine));
        }
        File.WriteAllText(this.filePath, builder.ToString());
    }

    public void Delete(int argLineNumber){
        string[] tempLines = File.ReadAllLines(this.filePath);
        StringBuilder builder = new StringBuilder();
        List<string> outLines = tempLines.ToList();
        outLines.RemoveAt(argLineNumber - 1);
        foreach (string tempLine in outLines) {
            builder.Append(String.Format("{0}\n", tempLine));
        }
        File.WriteAllText(this.filePath, builder.ToString());
    }

    public void Write(string argLine){
        string[] tempLines = File.ReadAllLines(this.filePath);
        string[] outLines = new string[tempLines.Length + 1];
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < tempLines.Length - 1; i++)
        {
            outLines[i] = tempLines[i];
        }
        outLines[tempLines.Length - 1] = argLine;
        outLines[tempLines.Length] = tempLines[tempLines.Length - 1];
        foreach (string tempLine in outLines) {
            builder.Append(String.Format("{0}\n", tempLine));
        }
        File.WriteAllText(this.filePath, builder.ToString());
    }
    
    public void Clear(){
        StringBuilder builder = new StringBuilder();
        builder.Append(String.Format("{0}\n", this.fileStart));
        builder.Append(String.Format("{0}\n", this.fileEnd));
        File.WriteAllText(this.filePath, builder.ToString());
    }

    public int Find(string argLine){
        int tempFind = -1;
        string[] tempLines = File.ReadAllLines(this.filePath);
        for (int i = 0; i < tempLines.Length; i++)
        {
            if (argLine.Equals(tempLines[i])) {
                tempFind = i + 1;
            }
        }
        return tempFind;
    }

    public int Size(){
        string[] tempLines = File.ReadAllLines(this.filePath);
        return tempLines.Length;
    }

    public void Copy(BaseFileHandler argIn){
        this.Clear();
        for (int i = 2; i < argIn.Size(); i++) {
            this.Insert(this.Size(), this.formData(argIn.extractData(argIn.Read(i))));
         }
    }

    abstract public string formData(Contact argIn);
    
    abstract public Contact extractData(string argIn);
}